storage/client_range_test.go

```diff
@@ -150,3 +154,158 @@ func TestRejectFutureCommand(t *testing.T) {
 		t.Errorf("expected 15, got %v", v)
 	}
 }
+
+// TestTxnPutOutOfOrder tests a case where a put operation of an older
+// timestamp comes after a put operation of a newer timestamp in a txn.
+//
+// 1) Writer executes a put operation with time T in a txn.
+// 2) Before the txn is committed, Reader sends a get
+//    operation with time T+100. This triggers the txn restart.
+// 3) Reader sends another get operation with time T+200. The
+//    write intent is resolved (and the txn timestamp is pushed to
+//    T+200), but the actual get operation has not yet been executed (and
+//    hence the timestamp cache has not been updated).
+// 4) Writer restarts the txn and executes the put operation
+//    again. The timestamp of the operation is T+100, which is earlier
+//    than the timestamp pushed at Step 3 (= T+200).
```

Comment:
Without reading the code below I'm still confused by this comment. It seems to mix current behavior with intended behavior. The test should only talk about intended behavior.